script/libraries/LibInitializeGuard.sol

```diff
@@ -0,0 +1,321 @@
+          || (lastInitVer == MAX_VER_V5 && $cache._initSlot[proxy].nBit == N_BIT_INIT_V5),
+        string.concat("LibInitializeGuard: Logic ", vm.getLabel(logics[i]), " does not disable initialized version!")
+      );
+    }
+  }
+
+  /**
+   * @dev Validate the initialized version of the proxies.
+   * - Check if `_initialized` slot is found.
+   * - Check if the last initialized version is equal to the number of `initialize` functions.
+   */
+  function _validateProxiesVersion(Cache storage $cache) private {
+    address[] memory proxies = $cache._proxies.values();
+    uint256 length = proxies.length;
+
+    for (uint256 i; i < length; ++i) {
+      address proxy = proxies[i];
+      InitializedSlot memory slot = $cache._initSlot[proxy];
+
+      require(
+        slot.found,
+        string.concat("LibInitializeGuard: Proxy ", vm.getLabel(proxies[i]), " does not have `_initialized` slot!")
+      );
+
+      uint256 lastInitVer = $cache._lastInitVer[proxy];
+
+      require(
+        lastInitVer != 0, string.concat("LibInitializeGuard: Proxy ", vm.getLabel(proxy), " does not initialize!")
+      );
+
+      if (
+        (lastInitVer == MAX_VER_V4 && slot.nBit == N_BIT_INIT_V4)
+          || (lastInitVer == MAX_VER_V5 && slot.nBit == N_BIT_INIT_V5)
+      ) {
+        string memory ret = vm.prompt(
+          string.concat(
+            "[WARNING] ".yellow(),
+            vm.getLabel(proxy),
+            " disabled initialized version.".yellow(),
+            " Is it intentional?\n".yellow(),
+            "Press ",
+            "yes ".blue(),
+            "to continue..."
+          )
+        );
+        require(keccak256(bytes(vm.toLowercase(ret))) == keccak256("yes"), "LibInitializeGuard: User aborted!");
+
+        continue;
+      }
+
+      uint256 initFnCount = _getInitializeFnCount($cache, proxy);
+      require(
+        lastInitVer == initFnCount,
+        string.concat(
+          "LibInitializeGuard: Invalid initialized version!",
+          " Expected: ",
+          vm.toString(initFnCount),
+          " Got: ",
+          vm.toString(lastInitVer)
+        )
+      );
+    }
+  }
+
+  /**
+   * @dev Validate the intermediate changes of the `_initialized` slot of the given `access` storage.
+   *
+   * @param acc The storage access of data.
+   * @param slot The initialized slot of the proxy.
+   * @return shouldSkip Whether to skip the validation.
+   */
+  function _validateInitChanges(Vm.StorageAccess memory acc, InitializedSlot memory slot)
+    private
+    view
+    returns (bool shouldSkip)
+  {
+    uint256 mask = (1 << slot.nBit) - 1;
+
+    uint256 prvVer = (uint256(acc.previousValue) >> slot.bitOffset) & mask;
+    uint256 newVer = (uint256(acc.newValue) >> slot.bitOffset) & mask;
+
+    // Skip if `_initialized` bytes location in `slot` does not change
+    // Assume other data in given slot is not related to initialized version
+    if (prvVer == newVer) return true;
```

Comment:
Should not allow overriding `_initialized` slot with the same value. It's kinda weird behavior.

Reply:
slot where `_initialized` is stored may have different data, such as address `_owner` sinced it is packed together, if we inherit `Ownable` extension, if in that case, we have to simply skip the check.

Reply:
Understood 🚀 rocket

Reply:
Agree with the case `_initialized` slot can be packed. However, when the `_initialized` slot is custom slot and could not be packed, `prvVer` should not equal to `newVer` in the this case